### PR TITLE
Async loading for outdir and proc-macro

### PR DESCRIPTION
**Pull request by edwin0cheng**
cc #7328

Peek 2021-01-24 02-04

[Edit] 
~~Finding a way to know when the workspace and build data are loaded...~~

[Edit 2] 
Not perfect solution, but seem to work now.

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Updated, use `OpQueue` to make sure only one task is running.

File: crates/rust-analyzer/tests/rust-analyzer/support.rs [hunk 1]
```diff
@@ -233,6 +233,70 @@ impl Server {
         });
         self
     }
+
+    pub(crate) fn wait_until_workspace_and_build_data_are_loaded(self) -> Server {
```
Comment (edwin0cheng) [inline crates/rust-analyzer/tests/rust-analyzer/support.rs:237]: The logic is quite complicated. But the problem here is our CI is not good on multi threading :( When priming caches, it eats a lot of cpu and coming requests will be too slow to run. 

The second problem here is , the ordering of the "roots_scan" events are based on how threads are run, such that is indeterminate. We tried out best here to guess when the actual "roots scan" ends (And we will have 2 roots scan event here)

**Comment by matklad** [comment]
Finally took a look here!

I think this is right in general terms, but there are a couple of concurrency-related things which make my old scars tingle :-)

First, I am unsure about

```
                if !build_data_loader.is_empty() {
                    sender
                        .send(Task::FetchBuildData(BuildDataProgress::Prepare(build_data_loader)))
                        .unwrap();
                }
```

This happens in the background thread which fetches the workspace. This I think
is prone to a race condition -- we might fetch build data before we actually
switched to the new workspace in the main loop. Its better to schedule build
data fetching on the main loop (from `&mut GlobalState`), after we switched the
workspace itself.

The second thing I think can cause pain is this mutation

```
        if let Some(build_data_loader) = &self.build_data_loader {
            for workspace in workspaces.iter_mut() {
                workspace.update_build_data(build_data_loader);
            }
        }
```

Fundamentally, what we have here are two source of information: workspace and build info.

The current code merges them by destructively updating the state. This is a lossy operation, which can not be trivially undone.
A more obviously correct approach is to preserve both pieces of data as is, and merge them on the fly, when swithcing the workspace.

I think we can do the following:

```
struct GlobalStatea {
    /// Invariant: `workspaces.len() == workspace_build_data.len()`
    pub(crate) workspaces: Arc<Vec<ProjectWorkspace>>,
    pub(crate) workspace_build_data: Vec<Option<BuildData>>,
    pub(crate) fetch_workspaces_queue: OpQueue,
}
```

Loading ws will set the first field, loading ws data will set the second one. When either of fields changes, we call `switch_workspaces` which sets the crategraph.

**Comment by edwin0cheng** [comment]
Fixed and sorry for the larger change set of this PR. 

It is due to the fact that I have to decouple the build data related config out of `CargoWorkspace` to make it works.

File: crates/rust-analyzer/tests/rust-analyzer/support.rs (hunk 1 above)
Comment (matklad) [reply crates/rust-analyzer/tests/rust-analyzer/support.rs:237]: For cache priming, I think we might want to just disable it for tests? 

For the second issue, I think this logic belongs to the main loop. Specifically (and not only in tests), rust-analyzer should be able to tell the clients: "the project is fully loaded and I can serve every request". 

We generally use the `status` field to track this event:

https://github.com/rust-analyzer/rust-analyzer/blob/ad3cb2125dee4379c11f447216f510d6544c10c5/crates/rust-analyzer/src/main_loop.rs#L381-L395

And we communicate this to client via the status notification:

https://github.com/rust-analyzer/rust-analyzer/blob/ad3cb2125dee4379c11f447216f510d6544c10c5/crates/rust-analyzer/src/reload.rs#L96-L99

I am not sure why we even use progress rathe than status in support, I am trying to fix this right now.

For this PR,I think we want to change the status logic to only become ready once external resources are loaded

**Comment by edwin0cheng** [comment]
> For this PR,I think we want to change the status logic to only become ready once external resources are loaded

On of the problem here is, the `Ready` transition is set by `vfs::loader::Message::Progress`: 

https://github.com/rust-analyzer/rust-analyzer/blob/ce2f0bdac726a34627d5cf14362e7bf14f868d09/crates/rust-analyzer/src/main_loop.rs#L279-L301

So do you want me to do something like this ? : 

```rust
vfs::loader::Message::Progress { n_total, n_done } => {
    if n_total == 0 {
        self.transition(Status::Invalid);
    } else {
        let state = if n_done == 0 {
            self.transition(Status::Loading);
            Progress::Begin
        } else if n_done < n_total {
            Progress::Report
        } else {
            assert_eq!(n_done, n_total);        
            if self.config.load_out_dirs_from_check() && self.workspace_build_data.is_some() {
                self.transition(Status::Ready);
            }
            Progress::End
        };
        self.report_progress(
            "roots scanned",
            state,
            Some(format!("{}/{}", n_done, n_total)),
            Some(Progress::fraction(n_done, n_total)),
        )
    }
}
```

**Comment by edwin0cheng** [comment]
> For cache priming, I think we might want to just disable it for tests?

Oh, but if we don't do cache priming, will the first request get longer to complete ?

**Comment by matklad** [comment]
> Oh, but if we don't do cache priming, will the first request get longer to complete ? 

for tests, I would expect the opposite, as we'll be doing only necessary work.

> One of the problem here is, the Ready transition is set by vfs::loader::Message::Progress but not by switch_workspaces:

Yeah, the specific transition rules (and the actual states we have) are tricky. I think we might want to split `Ready` into `Ready` and `ReadyPartial`, to properly capture the "we have the meta, but not the build stuff yet". I wonder though, whould two states be enough?

File: crates/rust-analyzer/tests/rust-analyzer/support.rs (hunk 1 above)
Comment (matklad) [reply crates/rust-analyzer/tests/rust-analyzer/support.rs:237]: >I am trying to fix this right now.

Hm, this is for some reason more complicated that I expect. For some reason, multitheraded tests are really slow even with --release...

File: crates/rust-analyzer/src/global_state.rs [hunk 2]
```diff
@@ -79,7 +82,9 @@ pub(crate) struct GlobalState {
     pub(crate) source_root_config: SourceRootConfig,
     pub(crate) proc_macro_client: Option<ProcMacroClient>,
     pub(crate) workspaces: Arc<Vec<ProjectWorkspace>>,
-    pub(crate) fetch_workspaces_queue: OpQueue,
+    pub(crate) fetch_workspaces_queue: OpQueue<Arc<Config>>,
```
Comment (matklad) [inline crates/rust-analyzer/src/global_state.rs:85]: Hm, I'd expect `OpQueue<()>` here.

```suggestion
    pub(crate) fetch_workspaces_queue: OpQueue<Arc<Config>>,
```

But I guess `Arc<Config>` would be semantically more appropriate.

File: crates/rust-analyzer/src/main_loop.rs [hunk 3]
```diff
@@ -372,13 +400,14 @@ impl GlobalState {
         }
 
         let state_changed = self.process_changes();
-        if prev_status == Status::Loading && self.status == Status::Ready {
+        let is_ready = self.status == Status::Ready || self.status == Status::ReadyPartial;
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop.rs:403]:
```suggestion
        let is_ready = matches!(self.status, Status::Ready | ReadyPartial);
```

File: crates/rust-analyzer/src/reload.rs [hunk 4]
```diff
@@ -99,18 +107,49 @@ impl GlobalState {
...
+        self.fetch_build_data_queue.request_op(build_data_collector);
+    }
+
+    pub(crate) fn fetch_build_data_if_needed(&mut self) {
+        let mut build_data_collector =
+            if let Some(build_data_collector) = self.fetch_build_data_queue.should_start_op() {
+                build_data_collector
+            } else {
+                return;
+            };
```
Comment (matklad) [inline crates/rust-analyzer/src/reload.rs:120]: match would be shorter here

File: crates/rust-analyzer/src/global_state.rs [hunk 5]
```diff
@@ -34,6 +36,7 @@ use crate::{
 pub(crate) enum Status {
     Loading,
     Ready,
+    ReadyPartial,
```
Comment (matklad) [inline crates/rust-analyzer/src/global_state.rs:39]: For the lsp_ext part, I think two separate enum variants makes sense. For this status however, I thing `Ready { partial: bool }` would work better?

**Review by matklad — COMMENTED** [review]
left a couple of nits, but, other that that, r=me

File: crates/rust-analyzer/src/global_state.rs (hunk 2 above)
Comment (edwin0cheng) [reply crates/rust-analyzer/src/global_state.rs]: I changed to () :)

File: crates/rust-analyzer/src/reload.rs (hunk 4 above)
Comment (edwin0cheng) [reply crates/rust-analyzer/src/reload.rs]: Fixed

File: crates/rust-analyzer/src/global_state.rs (hunk 5 above)
Comment (edwin0cheng) [reply crates/rust-analyzer/src/global_state.rs]: Fixed

File: crates/rust-analyzer/src/main_loop.rs (hunk 3 above)
Comment (edwin0cheng) [reply crates/rust-analyzer/src/main_loop.rs]: Fixed.

**Comment by edwin0cheng** [comment]
And I removed the special case `wait_until_workspace_and_build_data_are_loaded` and use `wait_until_workspace_is_loaded` directly.

**Comment by edwin0cheng** [comment]
bors r=@maklad

**Pull request merged**